### ISSUE TITLE
XOR training sample

### ISSUE THREAD

**Issue by FireDragonGameStudio** [migueldeicaza/TensorFlowSharp#230]
Hy, is there a simple XOR training sample? I tried to implement XOR following this TensorFlow tutorial [https://aimatters.wordpress.com/2016/01/16/solving-xor-with-a-neural-network-in-tensorflow/](https://aimatters.wordpress.com/2016/01/16/solving-xor-with-a-neural-network-in-tensorflow/) but I can't make it work. :(

Best regards :)

**Comment by Nurator** [comment]
I am at the exact same point in my learning routine. It would be really nice to have something basic like that that works :) Right now, I nearly got it running, here is my code: 

`
using (var session = new TFSession())
            {

                var graph = session.Graph;

                double[,] xData = new double[,]
                {
                    {0, 0},
                    {0, 1},
                    {1, 0},
                    {1, 1}
                };
                double[] yData = new double[]
                {
                    0,
                    1,
                    1,
                    0
                };
                var x = graph.Placeholder(TFDataType.Double, new TFShape(4,2));
                var y = graph.Placeholder(TFDataType.Double, new TFShape(4, 1));

                var Theta1 = graph.VariableV2(new TFShape(2, 2), TFDataType.Double);
                var Theta2 = graph.VariableV2(new TFShape(2, 1), TFDataType.Double);
                var Bias1 = graph.VariableV2(new TFShape(2), TFDataType.Double);
                var Bias2 = graph.VariableV2(new TFShape(1), TFDataType.Double);
                var A2 = graph.Sigmoid(graph.Add(graph.MatMul(x, Theta1), Bias1));
                var Hypothesis = graph.Sigmoid(graph.Add(graph.MatMul(A2, Theta2), Bias2));
                
                var initTheta1 = graph.Assign(Theta1, graph.RandomNormal(new TFShape(2, 2)));
                var initTheta2 = graph.Assign(Theta2, graph.RandomNormal(new TFShape(2, 1)));
                var initBias1 = graph.Assign(Bias1, graph.RandomNormal(new TFShape(2)));
                var initBias2 = graph.Assign(Bias2, graph.RandomNormal(new TFShape(1)));
                

                var ones = graph.OnesLike(y);
                var firstcost = graph.Mul(y, graph.Log(Hypothesis));
                var secondcost = graph.Mul(graph.Sub(ones, y), graph.Log(graph.Sub(ones, Hypothesis)));
                //var cost = graph.ReduceMean(graph.Neg(graph.Add(firstcost, secondcost)));
                var cost = graph.ReduceSum(graph.SquaredDifference(y, Hypothesis));
                //var cost = graph.ReduceMean(graph.Abs(graph.Sub(y, Hypothesis)));

                var learning_rate = 0.01;
                var grad = graph.AddGradients(new TFOutput[] { cost }, new TFOutput[] { Theta1, Theta2, Bias1, Bias2 });
                var optimize = new[]
                {
                    graph.AssignSub(Theta1,graph.Mul(grad[0],graph.Const(learning_rate))).Operation,
                    graph.AssignSub(Theta2,graph.Mul(grad[1],graph.Const(learning_rate))).Operation,
                    graph.AssignSub(Bias1,graph.Mul(grad[2],graph.Const(learning_rate))).Operation,
                    graph.AssignSub(Bias2,graph.Mul(grad[3],graph.Const(learning_rate))).Operation
                };
                session.GetRunner().AddTarget(initTheta1.Operation, initTheta2.Operation,initBias1.Operation,initBias2.Operation).Run();
                
                for (var i = 0; i < 100000; i++)
                {
                    var result = session.GetRunner()
                        .AddInput(x, xData)
                        .AddInput(y, yData)
                        .AddTarget(optimize)
                        .Fetch(cost, Hypothesis, Theta1, Theta2, Bias1, Bias2).Run();
                    
                    if (i % 10000 == 0)
                    {
                        double runCost = (double) result[0].GetValue();
                        double[,] HypoArray = (double[,])result[1].GetValue();
                        double[,] runTheta1 = (double[,])result[2].GetValue();
                        double[,] runTheta2 = (double[,])result[3].GetValue();
                        double[] runBias1 = (double[])result[4].GetValue();
                        double[] runBias2 = (double[])result[5].GetValue();
                        Console.WriteLine("Cost:");
                        Console.WriteLine(runCost);
                        Console.WriteLine("Hypothesis:");
                        for (var index = 0; index < 4; index++)
                        {
                            Console.WriteLine("y: {0} Hypothesis: {1}", yData[index], HypoArray[index, 0]);
                        }
                        Console.WriteLine("Theta1:");
                        for (var indexI = 0; indexI < 2; indexI++)
                        {
                            Console.WriteLine(runTheta1[indexI, 0] + " " + runTheta1[indexI,1]);                            
                        }
                        Console.WriteLine("Theta2:");
                        for (var indexI = 0; indexI < 2; indexI++)
                        {
                            Console.WriteLine(runTheta2[indexI, 0]);
                        }
                        Console.WriteLine("Bias1:");
                        for (var indexI = 0; indexI < 2; indexI++)
                        {
                            Console.WriteLine(runBias1[indexI]);
                        }
                        Console.WriteLine("Bias2:");
                        for (var indexI = 0; indexI < 1; indexI++)
                        {
                            Console.WriteLine(runBias2[indexI]);
                        }
                    }
                }
            }
`

Unfortunately, it always learns 0.5 for all inputs, so something is wrong... I have no idea, why. But maybe someone with more understanding can use this code, fix the issue and then use it as a tutorial for future users.
